compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -3616,8 +3638,24 @@ private boolean isAssignableToUnionType(BType source, BType target, Set<TypePair
                     break;
                 }
 
-                if (isAssignable(sourceMember, targetMember, unresolvedTypes)) {
-                    sourceTypeIsNotAssignableToAnyTargetType = false;
+                if (sourceMember.tag == TypeTags.ERROR) {
+                    if (targetUnionIsDistinctError) {
+                        sourceTypeIsNotAssignableToAnyTargetType = !isErrorTypeAssignable((BErrorType) sourceMember,
+                                (BErrorType) targetMember, unresolvedTypes);
+                    } else if (allTargetTypesErrors && targetMember.tag == TypeTags.ERROR) {
+                        sourceTypeIsNotAssignableToAnyTargetType =
+                                !isErrorTypeAssignableWithNoTypeIds((BErrorType) sourceMember,
+                                        (BErrorType) targetMember, unresolvedTypes);
+                    } else {
+                        sourceTypeIsNotAssignableToAnyTargetType = !isAssignable(sourceMember, targetMember,
+                                unresolvedTypes);
+                    }
+                } else {
+                    sourceTypeIsNotAssignableToAnyTargetType = !isAssignable(sourceMember, targetMember,
+                            unresolvedTypes);
+                }
+
+                if (!sourceTypeIsNotAssignableToAnyTargetType) {
                     break;
```

Comment:
If we break here we are not agreeing with original logic. Here we are inside a while loop which goes through targets.
Then we are breaking when a source type is not assignable to one of the targets.

Reply:
Actually, we are breaking here for the negation of source type is not assignable to one of the targets

Reply:
But we are still going through a loop to check remaining sources against remaining targets, right? Shouldn't we check this after the loop?

Reply:
But we are returning early after `!sourceTypeIsNotAssignableToAnyTargetType` right?

Reply:
Check the inner loop in line 3610. I believe this logic should be around line 3580